text/0004-c-conversion-conventions.md

```diff
@@ -0,0 +1,281 @@
+
+# Constructing/Destructing Objects
+The rules for constructing new instances of types depend on their kind.
+
+## 1. opaqueptr
+This kind is constructed simply by calling new and delete.
+```c++
+// C++
+namespace FOO {
+struct Bar {
+    Bar(int a);
+    ~Bar();
+} CPPMM_OPAQUEPTR;
+}
+
+// C
+FOO_Bar* FOO_Bar_new(int a) {
+    return to_c(new FOO::Bar(a));
+}
+
+void FOO_Bar_delete(FOO_Bar* self) {
+    delete to_cpp(self);
+}
+```
+
+## 2. opaquebytes and valutype
+Since the caller allocates the memory, these are constructed using placement new, with the caller passing in a pointer to the pre-allocated region. Destructing means calling the destructor explicitly and the caller is responsible for freeing the associated memory.
+```c++
+// C++
+namespace FOO {
+struct Bar {
+    Bar(int a);
+    ~Bar();
+} CPPMM_OPAQUEBYTES;
+}
+
+// C
+FOO_Bar* FOO_Bar_ctor(FOO_Bar* self, int a) {
+    return to_c( new (self) FOO::Bar(a));
+}
+
+void FOO_Bar_dtor(FOO_Bar* self) {
+    to_cpp(self)->~Bar();
+}
+```
+
+# Naming Rules
+
+## 1. prepend entire namesace hierarchy to declaration name
+Consider the following C++ code
+```c++
+namespace FOO {
+struct Bar {
+    enum BazThings { Thing1, Thing2 };
+    void baz();
+};
+
+void qux();
+}
+```
+
+This would generate the following C declarations:
+```c
+typedef struct FOO_Bar FOO_Bar;
+
+enum FOO_Bar_BazThings { FOO_Bar_BazThings_Thing1, FOO_Bar_BazThings_Thing2 };
+
+void FOO_Bar_baz(FOO_Bar* self);
+
+void FOO_qux();
+```
+
+### Considerations
+Most (all?) of the libraries we're interested in version their namespaces to avoid dependency hell. cppmm currently allows renaming namespaces with the `-n` flag to make nicer names, for instance:
+```c
+OIIO_ImageInput_open()
+```
+instead of:
+```c
+OpenImageIO_v2_8_ImageInput_open()
+```
+though this could also be shortened to:
+```c
+OIIO28_ImageInput_open()
+```
+or somesuch.
+
+Unfortunately I think we might have to bake the version into the C API "namespaces" for exactly the same reason as in C++. Consider a Rust crate that transitively depends on two different versions of OIIO - might we get symbol clashing here? I'm not sure... pretty sure you would if the binding stubs were compiled as shared libraries, maybe not if they're static?
+
+## 2. The "self" pointer in methods should be called "self"
+Obviously `this` is taken and Rust and Python both use `self` so it should be familiar to everyone reading.
+
+## 3. Constructors/destructors
+### opaqueptr
+Just use `new` and `delete`
+```c
+FOO_Bar* FOO_Bar_new();
+FOO_Bar_delete(const FOO_Bar* self);
+```
+
+### opaquebytes and valuetype
+Use `ctor` and `dtor` - these are short, unlikely to clash, and describe exactly what's going on:
```

Comment:
Just throwing it out there. Could the naming be part of the type name instead of the constructor/destructor?
Like,
BFoo for opaque bytes.
VFoo for value.
OFoo for opaque.

Then you keep using new/delete, but it's clear than the OFoo is a wrapper around foo that is opaque and heap allocated?

Reply:
This is an interesting idea. It’s a neat solution. The downside is that the vast majority of functions *will* be called the same way between all the kinds, and we’d be changing all of them in a surprising way for the sake of making two of them look neater. 

Once I’ve got further with the refactor I’ll add this as an option so we can try it out and see how it looks.